# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/windows/sdn/sdnDocAccessible.cpp
# Commit: a53fb045a8e3
# Full Hash: a53fb045a8e34c1e1f14ca7b705eb9a5a0e76db2
# Author: James Teh <jteh@mozilla.com>
# Date: 2021-05-02 09:29:52
# Regressor Bug: 1694865
# File Overlap Count: 1
# Description:
#   Bug 1694865 part 19: Make EnumVariant, GeckoCustom and sdn*Accessible classes aggregate MsaaAccessible/MsaaDocAccessible instead of AccessibleWrap/DocAccessibleWrap. r=morgan
#   
#   These objects need to aggregate the primary COM object (MsaaAccessible).
#   Once the IUnknown implementation moves out of AccessibleWrap, it won't be possible to aggregate *AccessibleWrap any more.
#   
# ==============================================================================

diff -r 5bd374b642fe -r a53fb045a8e3 accessible/windows/sdn/sdnDocAccessible.cpp
--- a/accessible/windows/sdn/sdnDocAccessible.cpp	Sat May 01 21:09:04 2021 +0000
+++ b/accessible/windows/sdn/sdnDocAccessible.cpp	Sat May 01 22:29:25 2021 +0000
@@ -20,17 +20,18 @@
 
 IMPL_IUNKNOWN_QUERY_HEAD(sdnDocAccessible)
 IMPL_IUNKNOWN_QUERY_IFACE(ISimpleDOMDocument)
-IMPL_IUNKNOWN_QUERY_TAIL_AGGREGATED(mAccessible)
+IMPL_IUNKNOWN_QUERY_TAIL_AGGREGATED(mMsaa)
 
 STDMETHODIMP
 sdnDocAccessible::get_URL(BSTR __RPC_FAR* aURL) {
   if (!aURL) return E_INVALIDARG;
   *aURL = nullptr;
 
-  if (mAccessible->IsDefunct()) return CO_E_OBJNOTCONNECTED;
+  DocAccessible* acc = mMsaa->DocAcc();
+  if (!acc) return CO_E_OBJNOTCONNECTED;
 
   nsAutoString URL;
-  mAccessible->URL(URL);
+  acc->URL(URL);
   if (URL.IsEmpty()) return S_FALSE;
 
   *aURL = ::SysAllocStringLen(URL.get(), URL.Length());
@@ -42,10 +43,11 @@
   if (!aTitle) return E_INVALIDARG;
   *aTitle = nullptr;
 
-  if (mAccessible->IsDefunct()) return CO_E_OBJNOTCONNECTED;
+  DocAccessible* acc = mMsaa->DocAcc();
+  if (!acc) return CO_E_OBJNOTCONNECTED;
 
   nsAutoString title;
-  mAccessible->Title(title);
+  acc->Title(title);
   *aTitle = ::SysAllocStringLen(title.get(), title.Length());
   return *aTitle ? S_OK : E_OUTOFMEMORY;
 }
@@ -55,10 +57,11 @@
   if (!aMimeType) return E_INVALIDARG;
   *aMimeType = nullptr;
 
-  if (mAccessible->IsDefunct()) return CO_E_OBJNOTCONNECTED;
+  DocAccessible* acc = mMsaa->DocAcc();
+  if (!acc) return CO_E_OBJNOTCONNECTED;
 
   nsAutoString mimeType;
-  mAccessible->MimeType(mimeType);
+  acc->MimeType(mimeType);
   if (mimeType.IsEmpty()) return S_FALSE;
 
   *aMimeType = ::SysAllocStringLen(mimeType.get(), mimeType.Length());
@@ -70,10 +73,11 @@
   if (!aDocType) return E_INVALIDARG;
   *aDocType = nullptr;
 
-  if (mAccessible->IsDefunct()) return CO_E_OBJNOTCONNECTED;
+  DocAccessible* acc = mMsaa->DocAcc();
+  if (!acc) return CO_E_OBJNOTCONNECTED;
 
   nsAutoString docType;
-  mAccessible->DocType(docType);
+  acc->DocType(docType);
   if (docType.IsEmpty()) return S_FALSE;
 
   *aDocType = ::SysAllocStringLen(docType.get(), docType.Length());
@@ -86,7 +90,7 @@
   if (!aNameSpaceURI) return E_INVALIDARG;
   *aNameSpaceURI = nullptr;
 
-  if (mAccessible->IsDefunct()) return CO_E_OBJNOTCONNECTED;
+  if (!mMsaa->DocAcc()) return CO_E_OBJNOTCONNECTED;
 
   if (aNameSpaceID < 0) return E_INVALIDARG;  // -1 is kNameSpaceID_Unknown
 
@@ -109,5 +113,5 @@
   if (!aCommaSeparatedMediaTypes) return E_INVALIDARG;
   *aCommaSeparatedMediaTypes = nullptr;
 
-  return mAccessible->IsDefunct() ? CO_E_OBJNOTCONNECTED : E_NOTIMPL;
+  return !mMsaa->DocAcc() ? CO_E_OBJNOTCONNECTED : E_NOTIMPL;
 }